src/include/duckdb/common/types.hpp

```diff
@@ -348,6 +348,9 @@ struct LogicalType {
 	bool Contains(F &&predicate) const;
 	bool Contains(LogicalTypeId type_id) const;
 
+	template<class F>
+	static LogicalType VisitReplace(const LogicalType &type, F &&func);
```

Comment:
nit: can we turn this into std::function, or maybe just move it to a separate class altogether (`LogicalTypeVisitor`)?